### PR TITLE
fixed GenericRow compare for different _fieldToValueMap size

### LINKED ISSUE #6963: GenericRow.compareMap always returns true for maps of different sizes
I believe this was a subtle bug introduced in #6525. The current code structure is
```
if (thisMap.size() == thatMap.size()) {
  ...
}
return true
```
If a GenericRow has a different number of fields, this always returns true. I don't believe this was the intention in that change. Previously it was doing `return _fieldToValueMap.equals(that._fieldToValueMap) && _nullValueFields.equals(that._nullValueFields);` which returns false on size differences.

### PR DESCRIPTION
**Pull request by jadami10**
## Description

fixes #6963 

GenericRow.compareMap will now return false if the maps are different sizes.

I'm fairly new to this project, so I do not believe this affects upgrades, but I am also not sure where GenericRow equality checks are done or how important they are.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? No

Does this PR fix a zero-downtime upgrade introduced earlier? No

Does this PR otherwise need attention when creating release notes? No

### CODE REVIEW

**Review by Jackie-Jiang — APPROVED** [review]
LGTM, good catch

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java [hunk 1]
```diff
@@ -227,26 +227,28 @@ public boolean equals(Object obj) {
...
-          if (fieldValue instanceof Map && thatFieldValue instanceof Map) {
-            return compareMap((Map<String, Object>) fieldValue, (Map<String, Object>) thatFieldValue);
-          }
-          if ((fieldValue instanceof byte[]) && (thatFieldValue instanceof byte[])) {
-            return Arrays.equals((byte[]) fieldValue, (byte[]) thatFieldValue);
-          }
-          if (fieldValue.getClass().isArray() && thatFieldValue.getClass().isArray()) {
-            return compareArray((Object[]) fieldValue, (Object[]) thatFieldValue);
-          }
+    if (thisMap.size() != thatMap.size()) {
```
Comment (xiangfu0) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:230]: nit: shall we do null check first?

Comment (jadami10) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:230]: is it possible for these to be null? If os, what would the desired behavior be here? Only return true if they're both null?

File: pinot-spi/src/test/java/org/apache/pinot/spi/data/readers/GenericRowTest.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.data.readers;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+public class GenericRowTest {
```
[the remark this answers is not in the document]
Comment (jadami10) [reply pinot-spi/src/test/java/org/apache/pinot/spi/data/readers/GenericRowTest.java:25]: Good idea, added a few more tests.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:230]: Right. Maybe reuse this helper method once it’s merged ?
https://github.com/apache/incubator-pinot/pull/6965/files#diff-a5239a99ae5c9433e0bc5fdd5d7b67e45e0f90607d9d21821d24f334bda27ca3R92

Comment (jadami10) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:230]: Oh, that looks perfect and I can totally remove `compareMap`. What do you think the ETA to merge 6595 is?

Comment (xiangfu0) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:230]: it's merged already, please ensure that method can handle nested objects.

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java [hunk 3]
```diff
@@ -217,65 +217,11 @@ public boolean equals(Object obj) {
     }
     if (obj instanceof GenericRow) {
       GenericRow that = (GenericRow) obj;
-      if (!_nullValueFields.containsAll(that._nullValueFields) || !that._nullValueFields
```
Comment (jadami10) [inline pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:220]: I'm not sure why #6525 expanded this, but two .containsAll should be the same as .equals

**Review by jadami10 — COMMENTED** [review]
This should hopefully be a good cleanup. It folds the special functionality of GenericRow's compareMap and compareArray back into EqualityUtils. I've also added a number of tests.

**Comment by jadami10** [comment]
friendly ping here. I know this was initially accepted, but I made some big changes.

I also need another approval for the review workflows to run

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:220]: make sense

**Comment by xiangfu0** [comment]
> friendly ping here. I know this was initially accepted, but I made some big changes.
> 
> I also need another approval for the review workflows to run

Can you run `mvn spotless:apply` to fix the styling?

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:220]: I think the issue was there could be a deep comparison for nested structure.

**Comment by jadami10** [comment]
> Can you run mvn spotless:apply to fix the styling?

done, sorry keep missing that step. can you approve the workflow one more time

File: pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java (hunk 3 above)
Comment (jadami10) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:220]: you might have better insight than me then. would you prefer I leave it as as two `.containsAll`?

Comment (xiangfu0) [reply pinot-spi/src/main/java/org/apache/pinot/spi/data/readers/GenericRow.java:220]: I think you should be good if the test passed.

**Comment by xiangfu0** [comment]
Thanks for fixing this! @jadami10

**Pull request merged**